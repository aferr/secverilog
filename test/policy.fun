(declare-fun epol (Int) Bool)

(assert (= (epol 0) false))
(assert (forall ((x Int)) (implies (not (= x 0)) (epol x) )))

(declare-fun miss (Int Int Int) Bool)
(assert (forall ((x Int) (y Int) (z Int))
	(ite (and (not (= x 0)) (<= y z)) (miss x y z) (not (miss x y z)))
	))

(declare-fun Domain (Int) Label)
(assert
	(forall ((x Int) (y Int)) (implies (<= x y) (leq (Domain x) (Domain y)))))
; Bottom bit (i.e., even integers) is the valid bit, if its zero then treat as TOP
(assert (forall ((x Int)) (implies (= (mod x 2) 0) (= HIGH (Domain x)))))
