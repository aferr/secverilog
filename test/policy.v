module policy(
	      input 	  clk,
	      input 	  rst,
	      input 	  invalid,
	      input [4:0] missId,
	      input [3:0] newSpecId,
	      input 	  newSpecValid,	      
	      output 	  {H} out
);


   reg 		    seq {erase(L;epol invalid;();H)} data_reg;
   reg 		    seq {erase(L;epol invalid;();H)} dreg_2;   

   
   always@(posedge clk) begin
      if (rst) data_reg <= 0;
      else data_reg <= data_reg; //should fail because policy may be true this cycle and become false next cycle
   end

   always@(posedge clk) begin
      if (rst || invalid) dreg_2 <= 0;
      else dreg_2 <= dreg_2; //should succeed because we don't copy if data was invalid
   end

   assign out = (invalid) ? data_reg : 0;

   reg [4:0] seq {erase(Domain spec;miss invalid, missId; spec;H)} spec;

   wire      com {H} inv = (invalid && missId <= spec) ? 1 : 0;   
   
   always@(posedge clk) begin
      if (rst)
	spec <= 0;
      else if (inv)     
       	spec <= 0; //should succeed since this resets it to a HIGH label for next cycle
      else
	spec <= spec; //should succeed since we've proven it not invalid
   end
endmodule
