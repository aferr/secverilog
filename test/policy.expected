unsat
unsat
Start base conditions are satisfiable? (should be sat)
sat
End base conditions are satisfiable?
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
sat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
unsat
